[WIP] Shibuya: dApps Staking 3.1.0 - fixed era lengths #552

jakehemmerle opened the pull request:
Check "Required Tasks" in  #552 to see what this currently includes

**Check list**
- [ ] contains breaking changes
- [ ] adds new feature
- [x] modifies existing feature (bug fix or improvements)
- [ ] updated spec version
- [ ] relies on other tasks
- [ ] documentation changes
- [x] tests and/or benchmarks are included
- [ ] changed API client type definition or chain metadata

Dinonard commented on frame/dapps-staking/src/pallet/mod.rs:
```diff
@@ -235,15 +240,18 @@ pub mod pallet {
 
             // Value is compared to 1 since genesis block is ignored
-            if now % blocks_per_era == BlockNumberFor::<T>::from(1u32)
+            if now == next_era_starting_block
```
This is good!

About the question for runtime upgrade - try to do the following:
* run a local node without your code
* build a new runtime which includes your code
* do runtime upgrade
* check what happens with era changes

jakehemmerle replied:
This seems quite obvious now that you've written the steps out :)

Dinonard commented on frame/dapps-staking/src/tests.rs:
```diff
@@ -1264,6 +1264,50 @@ fn new_era_forcing() {
+        // start new era era and record era number and height
+        advance_to_era(mock::DappsStaking::current_era() + 1);
+        let mut expected_era = mock::DappsStaking::current_era();
+        let mut previous_era_starting_block_height;
```
I think you should avoid such empty declarations - instead declare the variable when you need it.
And please remove `height` from name, it's very confusing :D 

Also, if you need to change the value of the variable multiple times, I'd suggest you just re-define it.
E.g.
```
let x = 5;
let x = 7;
```

jakehemmerle replied:
Sorry for the confusion! Perhaps I'm using it wrong. I had a ways heard of 'block height' defined as: "The block height of a particular block is defined as the number of blocks preceding it in the blockchain".

I'll change it to 'block_number".

Dinonard replied:
I didn't know that!

I did a check in substrate repo and block height is mentioned a few times.
Block number is mentioned much more though.

For our repo, I suggest we stick with block number since we use that term already in the code (we don't use block height).

Thanks for the explanation though, I learned something :D

Dinonard commented on frame/dapps-staking/src/tests.rs:
```diff
@@ -1264,6 +1264,50 @@ fn new_era_forcing() {
+
+
+        // now, move to middle of an era, force new era and assert the next block starts a new era
+        run_for_blocks(1);
```
Perhaps make this dependent on the blocks per era constant?

E.g. if BLOCKS_PER_ERA gets modified in the future, you'd want your UT to still work.

jakehemmerle replied:
I would have usually done some value related to division of blocks_per_era, but I decided to hardcode this since blocks_per_era for the local runtime was only 3 blocks, which is the smallest era value to realistically test this (although technically 2 could work). I don't expect this to change Down to 2, and the test won't break as long as blocks_per_era >= 3.

Dinonard commented on frame/dapps-staking/src/tests.rs:
```diff
@@ -1264,6 +1264,50 @@ fn new_era_forcing() {
+        previous_era_starting_block_height = current_era_starting_block_height;
+        current_era_starting_block_height = System::block_number();
+        assert_eq!(mock::DappsStaking::current_era(), expected_era);
+        assert_eq!(current_era_starting_block_height - previous_era_starting_block_height, blocks_per_era);
```
Two pieces of advice:
1) I'd suggest you break this UT into two separate tests: one for normal run and one that uses `force new era`. Makes it more clear what's being tested and easier to maintain in the future.
2) Reusing the same variable many times seems convenient but you risk a higher chance of making a mistake somewhere. In this particular scenario, you could end up modifying one variable (previous_era_starting_block) but forget about another (current_era_starting_block).

jakehemmerle replied:
1. Sure, I can split them into 2 tests.

2. I deviated from the "don't reuse a variable" rule because I thought it significantly increased readability (and therefore reliability) since pattern happens 3 times times in this test:

```rust
advance_to_era(mock::DappsStaking::current_era() + 1); // or something that should increases the era
expected_era += 1;
previous_era_starting_block_height = current_era_starting_block_height;
current_era_starting_block_height = System::block_number(); // update block height
assert_eq!(mock::DappsStaking::current_era(), expected_era); // make sure line 1 of code increased the current era
assert_eq!(current_era_starting_block_height - previous_era_starting_block_height, blocks_per_era); // make sure the distance between the blocks is what we expected
```

Readability was also same reason I left `previous_era_starting_block_height` uninitialized until line 1282. Makes the above pattern easier for the eyes to pick up on.

I originally was not reusing any variables, but readability became poorer as I had to find weird or super long variable names for storing and asserting the block numbers.

Dinonard reviewed (commented):
Good work!

I left some comments and please don't hesitate to reply/ask questions/disagree.

jakehemmerle commented:
Next:
- `pallet-dapps-staking` v2.1.0 to get merged to `polkadot-v0.9.13` and tagged.
- update local, shibuya, and shiden runtime dependencies with new `astar-frame` repo tags
- upgrade Astar runtime `pallet-dapps-staking` from v1.1.3 to v2.0.0, then to this with 2.1.0

Dinonard commented on runtime/shibuya/src/lib.rs:
```diff
@@ -751,8 +751,46 @@ pub type Executive = frame_executive::Executive<
+        use pallet_dapps_staking::{Config as DappStakingConfig, CurrentEra, NextEraStartingBlock};
+        use sp_runtime::SaturatedConversion;
+
+        // We can always use this formula to calculate the offset to compute the next era starting block
```
Please clean up surplus comments here and in Shiden code.

Dinonard commented on runtime/shibuya/src/lib.rs:
````diff
@@ -751,8 +751,46 @@ pub type Executive = frame_executive::Executive<
+        // let DAPP_STAKING_BLOCK_OFFSET = current_block - (current_block % blocks_per_era) - (current_block * blocks_per_era);
+        // ```
+
+        let dapps_staking_block_offset: u32 = 172800; // Shibuya: 819166 - (819166 % 1200) - (538 * 1200)
````
So this is the block at which era 1 started?

jakehemmerle replied:
Yes

jakehemmerle commented on runtime/shibuya/Cargo.toml:
```diff
@@ -76,8 +76,8 @@ polkadot-runtime-common = { git = "https://github.com/paritytech/polkadot", defa
-pallet-dapps-staking = { git = "https://github.com/AstarNetwork/astar-frame", tag = "pallet-dapps-staking-2.1.0/polkadot-v0.9.16", default-features = false }
-pallet-precompile-dapps-staking = { git = "https://github.com/AstarNetwork/astar-frame", tag = "pallet-precompile-dapps-staking-1.1.1/polkadot-v0.9.16", default-features = false }
+pallet-dapps-staking = { git = "https://github.com/AstarNetwork/astar-frame", branch = "fixed-era-length", default-features = false }
+pallet-precompile-dapps-staking = { git = "https://github.com/AstarNetwork/astar-frame", branch = "fixed-era-length", default-features = false }
```
fix deps after merge

Dinonard commented on runtime/shibuya/src/lib.rs:
```diff
@@ -798,8 +798,35 @@ pub type Executive = frame_executive::Executive<
+        NextEraStartingBlock::<Runtime>::put(
+            next_era_starting_block.saturated_into::<<Runtime as Config>::BlockNumber>(),
+        );
+        <Runtime as Config>::DbWeight::get().reads(1)
+            + <Runtime as Config>::DbWeight::get().writes(1)
```
You can use `<Runtime as Config>::DbWeight::get().reads_writes(1, 1)`

Dinonard commented on runtime/shibuya/src/lib.rs:
```diff
@@ -7,7 +7,7 @@
 use frame_support::{
     construct_runtime, parameter_types,
-    traits::{Contains, Currency, FindAuthor, Imbalance, Nothing, OnUnbalanced},
+    traits::{Contains, Currency, FindAuthor, Imbalance, Nothing, OnUnbalanced, OnRuntimeUpgrade},
```
Please make sure to run `cargo fmt` before push.
Or just push another commit if CI fails.

Dinonard reviewed (commented):
Looks ready for production :+1:, just sort out the two comments.

We should probably start with Shibuya first and then move to Shiden and Astar.
Let's discuss it in Slack.

jakehemmerle commented:
@Dinonard  So for testing Shibuya:

Firstly, I used `try-runtime`, which I used no hooks and the upgrade did not throw any errors or anything.

Then, using `polkadot-launch`, I created a local testnet with the latest runtime, tweaked this upgrade's block number offset to accommodate the local testnet block numbers, then compiled this updated shibuya runtime with release parameters and manually upgraded via `setCode()`. After upgrade, I confirmed that the next era block number was the number it was supposed to be, and watched the new runtime emit the new as it was supposed to.

I added an additional test to make sure the next era block is properly recalculated if we enable maintenance mode (the greater than functionality), and while the test works and performs as expected, I did not recompile and test this functionality with `polkadot-launch` after I added this test.

I did not do any testing with `substrate-simnode`

Dinonard commented:
> @Dinonard So for testing Shibuya:
> 
> Firstly, I used `try-runtime`, which I used no hooks and the upgrade did not throw any errors or anything.
> 
> Then, using `polkadot-launch`, I created a local testnet with the latest runtime, tweaked this upgrade's block number offset to accommodate the local testnet block numbers, then compiled this updated shibuya runtime with release parameters and manually upgraded via `setCode()`. After upgrade, I confirmed that the next era block number was the number it was supposed to be, and watched the new runtime emit the new as it was supposed to.
> 
> I added an additional test to make sure the next era block is properly recalculated if we enable maintenance mode (the greater than functionality), and while the test works and performs as expected, I did not recompile and test this functionality with `polkadot-launch` after I added this test.
> 
> I did not do any testing with `substrate-simnode`

Good! :+1: 

I recommend to usually write up the testing you did as part of the PR summary.
It's neat to see this at the top since it's immediately clear what's been done so far.

For `try-runtime`, I'd recommend using pre/after hooks since they are great way of checking whether everything went as expected after upgrade. But it's ok.

There are 3 more things you need to update prior to merge:
1) Used tag for `pallet-dapps-staking` and `pallet-precompiles-dapps-staking` (you already have a TODO for that so it's covered)
2) Spec version for Shibuya runtime - you need to increment it by 1
3) Minor release versions of astar-collator, local-runtime, shibuya-runtime, shiden-runtime & astar-runtime should be incremented by 1 (check some older commits which contained runtime upgrade to see exactly what need to be done)

After that, we should be ready to go.

Dinonard commented:
Also, what's with the test CI?

Could it be related to recent optimization @shunsukew ?

Dinonard commented on runtime/shibuya/src/lib.rs:
```diff
@@ -796,7 +796,7 @@ pub type Executive = frame_executive::Executive<
     Runtime,
     AllPalletsWithSystem,
-    (DappsStakingMigrationV3,),
+    (DappsStakingMigrationV3, DappsStakingFixEraLength),
```
We don't need V3 migration anymore - it might even mess something up.
Feel free to delete it.

Dinonard reviewed (approved):
LGTM
